File: src/main/java/fastandslowpointers/happynumber/HappyNumber.java
package fastandslowpointers.happynumber;

public class HappyNumber {

    public boolean isHappyNumber(int n) {
        var fast = sumDigits(n);
        while (fast != 1 && n != fast) {
            n = sumDigits(n);
            fast = sumDigits(sumDigits(fast));
        }
        return fast == 1;
    }

    private int sumDigits(int number) {
        int totalSum = 0;
        while (number != 0) {
            int digit = number % 10;
            number = number / 10;
            totalSum += (Math.pow(digit, 2));
        }
        return totalSum;
    }

}

File: src/main/kotlin/fastandslowpointers/happynumber/HappyNumber.kt
package fastandslowpointers.happynumber

fun isHappyNumber(n: Int): Boolean {

    fun Int.sumDigits() = toString().fold(0) { acc, c -> acc + c.digitToInt() * c.digitToInt() }

    var slow = n
    var fast = n.sumDigits()
    while (fast != 1 && fast != slow) {
        slow = slow.sumDigits()
        fast = fast.sumDigits().sumDigits()
    }
    return fast == 1
}
